### PR TITLE
Tools for extracting features from PE files

### PR DESCRIPTION
**Pull request by esl-lewis**
Committing the scripts used to extract characteristics from PE files, analyse compilation dates and encode into ML features. 

Additionally an initial decision tree learning model.

### CODE REVIEW

**Comment by sebastien-rettie** [comment]
Thanks a lot @esl-lewis! Would it be possible to remove the unnecessary commented code before merging?

Cheers.

Sébastien.

File: ml-models/decision-tree.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        ("Categorical", onehot_transform, selector(dtype_include="object")),
+    ],
+    remainder="passthrough",
+)
+
+preprocess = Pipeline(steps=[("preprocess", column_trans)])
+
+# Preprocessing
+x = preprocess.fit_transform(x)
+# It is bad practice to have split after transform- fix
```
Comment (sebastien-rettie) [inline ml-models/decision-tree.py]: Is the plan to fix this before merging, or to go ahead with this as a first step?

**Comment by esl-lewis** [comment]
> 
> 
> Thanks a lot @esl-lewis! Would it be possible to remove the unnecessary commented code before merging?
> 
> Cheers.
> 
> Sébastien.

👍  Have cleaned these up in latest commit.

File: ml-models/pe-decision-tree.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+)
+
+pipeline = Pipeline(steps=[("preprocess", column_trans)])
+
+# Preprocessing and vectorise
+X = pipeline.fit_transform(x)
+
+# todo: fix as is bad practice to have split after preprocessing
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=4)
```
Comment (sebastien-rettie) [inline ml-models/pe-decision-tree.py]: Is there a particular motivation for this test/train split?

**Comment by sebastien-rettie** [comment]
Thanks for the great work @esl-lewis! I'm wondering if it would be possible to modularize the code to avoid code duplication as much as possible in the various ml-models? E.g. the functions `generate_types` and `get_feature_out` etc. that are used by multiple models could be defined only once in a common file, that would then be imported by the various ml-models. What do you think?

**Comment by esl-lewis** [comment]
> 
> 
> Thanks for the great work @esl-lewis! I'm wondering if it would be possible to modularize the code to avoid code duplication as much as possible in the various ml-models? E.g. the functions `generate_types` and `get_feature_out` etc. that are used by multiple models could be defined only once in a common file, that would then be imported by the various ml-models. What do you think?

Agreed! Have extracted some functions in the latest commit.

**Comment by sebastien-rettie** [comment]
Fantastic work, thanks a lot @esl-lewis! I'm happy with all these updates. Could you add some instructions (either in the form of a README or something else) detailing the workflow? Ideally these instructions would allow newcomers to reproduce the nice results you've shown so far (last request I promise!).

**Comment by esl-lewis** [comment]
> 
> 
> Fantastic work, thanks a lot @esl-lewis! I'm happy with all these updates. Could you add some instructions (either in the form of a README or something else) detailing the workflow? Ideally these instructions would allow newcomers to reproduce the nice results you've shown so far (last request I promise!).

Added some documentation to the README, not sure if markdown is behaving so let me know if it turns out okay!

**Comment by sebastien-rettie** [comment]
Hi @esl-lewis thanks for the updates (and apologies for the slow turnaround)! I've created a small PR into your branch adding a few markdowns and fixing the image. Once you accept the PR into your branch I'll merge this branch into `main`.

**Review by sebastien-rettie — APPROVED** [review]

File: ml-models/decision-tree.py (hunk 1 above)
Comment (esl-lewis) [reply ml-models/decision-tree.py]: I think this is an old version of the script, it has been fixed in current version.

**Pull request merged**